File: 75 days Challenge Leetcode/BT_DFS/Day-34/Solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Helper function to collect all leaf nodes of a tree
    def getLeaves(self, root, leaves):
        if not root:
            return
        
        if not root.left and not root.right:  # Leaf node
            leaves.append(root.val)
            return

        self.getLeaves(root.left, leaves)
        self.getLeaves(root.right, leaves)

    def leafSimilar(self, root1, root2):
        leaves1 = []
        leaves2 = []
        self.getLeaves(root1, leaves1)
        self.getLeaves(root2, leaves2)
        return leaves1 == leaves2  # Compare leaf sequences

File: 75 days Challenge Leetcode/BT_DFS/Day-34/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.ArrayList;
 import java.util.List;


class TreeNode {
    int val; 
    TreeNode left; 
    TreeNode right; 

    TreeNode(int x) {
        val = x;
        left = null;
        right = null;
    }
}


 class Solution {
     // Helper function to collect all leaf nodes of a tree
     private void getLeaves(TreeNode root, List<Integer> leaves) {
         if (root == null) return;
         
         if (root.left == null && root.right == null) { // Leaf node
             leaves.add(root.val);
             return;
         }
 
         getLeaves(root.left, leaves);
         getLeaves(root.right, leaves);
     }
     
     public boolean leafSimilar(TreeNode root1, TreeNode root2) {
         List<Integer> leaves1 = new ArrayList<>();
         List<Integer> leaves2 = new ArrayList<>();
         getLeaves(root1, leaves1);
         getLeaves(root2, leaves2);
         return leaves1.equals(leaves2); // Compare leaf sequences
     }
 }
 